draft-ietf-quic-transport.md

```diff
@@ -1449,20 +1417,121 @@ of three ways:
++                                                               +
+|                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                        Random Octets (*)                    ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+
+This packet uses the short header form with the shortest possible packet number
+encoding.  This minimizes the perceived gap between the last packet that the
+server sent and this packet.  The leading octet of the Stateless Reset Token
+will be interpreted as a packet number.  A server MAY use a different short
+header type, indicating a different packet number length, but this allows for
+the message to be identified as a stateless reset more easily using heuristics.
+
+A server copies the connection ID field from the packet that triggers the
+stateless reset.  A server omits the connection ID if explicitly configured to
+do so, or if the client packet did not include a connection ID.
+
+After the first short header octet and optional connection ID, the server
+includes the value of the Stateless Reset Token that it included in its
+transport parameters.
+
+After the Stateless Reset Token, the server pads the message with random
+octets.
+
+This design ensures that a stateless reset packet is - to the extent possible -
+indistinguishable from a regular packet.
+
+
+### Detecting a Stateless Reset
+
+A client detects a potential stateless reset when a packet with a short header
+cannot be decrypted.  The client then performs a constant-time comparison of the
+16 octets that follow the Connection ID with the Stateless Reset Token provided
+by the server in its transport parameters.  If this comparison is successful,
+the connection MUST be terminated immediately.  Otherwise, the packet can be
+discarded.
+
+
+### Selecting a Stateless Reset Token
+
+In order to create a Stateless Reset Token, a server could randomly generate
+{{!RFC4086}} a secret for every connection that it creates.  However, this
+presents a coordination problem when there are multiple servers in a cluster or
+a storage problem for a server that might lose state.  Stateless reset
+specifically exists to handle the case where state is lost, so this approach is
+suboptimal.
+
+A single static key can be used across all connections to the same endpoint by
+generating the proof using a second iteration of a preimage-resistant function
+that takes three inputs: the static key, a the connection ID for the connection
+(see {{connection-id}}), and an identifier for the server instance.  A server
+could use HMAC {{?RFC2104}} (for example, HMAC(static_key, server_id ||
+connection_id)) or HKDF {{?RFC5869}} (for example, using the static key as input
+keying material, with server and connection identifiers as salt).  The output of
+this function is truncated to 16 octets to produce the Stateless Reset Token
+for that connection.
```

Comment:
A server could not use a connection id because the connection id might have changed, and even if didn't change, it might be available in a short header. It also would not be able to use IP and port tuple because it also might have changed.